src/Tools/AnalyzerRunner/Program.cs

```diff
@@ -0,0 +1,375 @@
+                {
+                    uniqueOutput.AppendLine(message);
+                }
+            }
+
+            string directoryName = Path.GetDirectoryName(fileName);
+            string fileNameWithoutExtension = Path.GetFileNameWithoutExtension(fileName);
+            string extension = Path.GetExtension(fileName);
+            string uniqueFileName = Path.Combine(directoryName, $"{fileNameWithoutExtension}-Unique{extension}");
+
+            File.WriteAllText(fileName, completeOutput.ToString(), Encoding.UTF8);
+            File.WriteAllText(uniqueFileName, uniqueOutput.ToString(), Encoding.UTF8);
+        }
+
+        private static Statistic GetSolutionStatistics(IEnumerable<Project> projects, CancellationToken cancellationToken)
+        {
+            ConcurrentBag<Statistic> sums = new ConcurrentBag<Statistic>();
+
+            Parallel.ForEach(projects.SelectMany(project => project.Documents), document =>
+            {
+                var documentStatistics = GetSolutionStatisticsAsync(document, cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+                sums.Add(documentStatistics);
+            });
+
+            Statistic sum = sums.Aggregate(new Statistic(0, 0, 0), (currentResult, value) => currentResult + value);
+            return sum;
+        }
+
+        // TODO consider removing this and using GetAnalysisResultAsync
+        // https://github.com/dotnet/roslyn/issues/23108
+        private static async Task<Statistic> GetSolutionStatisticsAsync(Document document, CancellationToken cancellationToken)
+        {
+            SyntaxTree tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+            SyntaxNode root = await tree.GetRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var tokensAndNodes = root.DescendantNodesAndTokensAndSelf(descendIntoTrivia: true);
+
+            int numberOfNodes = tokensAndNodes.Count(x => x.IsNode);
+            int numberOfTokens = tokensAndNodes.Count(x => x.IsToken);
+            int numberOfTrivia = root.DescendantTrivia(descendIntoTrivia: true).Count();
+
+            return new Statistic(numberOfNodes, numberOfTokens, numberOfTrivia);
+        }
+
+        private static IEnumerable<DiagnosticAnalyzer> FilterAnalyzers(IEnumerable<DiagnosticAnalyzer> analyzers, Options options)
+        {
+            foreach (var analyzer in analyzers)
+            {
+                if (options.UseAll)
+                {
+                    yield return analyzer;
+                }
+                else if (options.AnalyzerNames.Count == 0)
+                {
+                    if (analyzer.SupportedDiagnostics.Any(diagnosticDescriptor => diagnosticDescriptor.IsEnabledByDefault))
+                    {
+                        yield return analyzer;
+                    }
+                }
+                else if (options.AnalyzerNames.Contains(analyzer.GetType().Name))
```

Comment:
is it 100% match or indexof might better so that user can give part of analyzer name rather than fullname?

Reply:
added to [https://github.com/dotnet/roslyn/issues/23108](https://github.com/dotnet/roslyn/issues/23108)

---
In reply to: [150148775](https://github.com/dotnet/roslyn/pull/23087#discussion_r150148775) [](ancestors = 150148775)